test-e2e/lp.test.ts

```diff
@@ -0,0 +1,162 @@
+      l1Dai.address,
+      l2DAITokenBridge.address,
+      l2Dai.address,
+      optimismConfig.Proxy__OVM_L1CrossDomainMessenger,
+      l1Escrow.address,
+      ZERO_GAS_OPTS,
+    ])
+    await waitForTx(l1Escrow.approve(l1Dai.address, l1DAITokenBridge.address, ethers.constants.MaxUint256))
+    expect(l1DAITokenBridge.address).to.be.eq(futureL1DAITokenBridgeAddress, 'Predicted address of l1DAITokenBridge doesnt match actual address')
+    console.log('L1 DAI Deposit: ', l1DAITokenBridge.address)
+
+    await waitForTx(l2Dai.rely(l2DAITokenBridge.address, ZERO_GAS_OPTS))
+    await waitForTx(l2Dai.deny(l2Signer.address, ZERO_GAS_OPTS))
+    await waitForTx(l2DAITokenBridge.deny(l2Signer.address, ZERO_GAS_OPTS))
+
+    l2Messenger = await ethers.getContractAt(L2_MESSENGER_ABI, optimismConfig._L2_OVM_L2CrossDomainMessenger, l2Signer)
+    lp = await deployUsingFactory(l1Signer, await l1.getContractFactory('LiquidityProvider'), [
+      optimismConfig.Proxy__OVM_L1CrossDomainMessenger,
+      ZERO_GAS_OPTS,
+    ])
+    console.log('LP: ', lp.address)
+    await lp.registerL1Token(l1Dai.address, l1DAITokenBridge.address)
+    await lp.registerL2Token(l2Dai.address, l2DAITokenBridge.address)
+  })
+
+  async function deposit() {
+    console.log('Depositing...')
+    await waitForTx(l1Dai.connect(l1User).approve(l1DAITokenBridge.address, depositAmount))
+    await waitToRelayTxsToL2(
+      l1DAITokenBridge.connect(l1User).depositERC20(l1Dai.address, l2Dai.address, depositAmount, defaultGasLimit, '0x'),
+      watcher,
+    )
+
+    // check post-deposit state
+    const userL2Balance = await l2Dai.balanceOf(l2User.address)
+    expect(userL2Balance.toString()).to.be.eq(depositAmount)
+    const userL1Balance = await l1Dai.balanceOf(l1User.address)
+    expect(userL1Balance.toString()).to.be.eq(initialL1DaiAmount.sub(depositAmount))
+  }
+
+  async function withdraw() {
+    console.log('Withdrawing...')
+    const messageNonce = parseInt(await l2Messenger.messageNonce())
+    const extraData = new utils.AbiCoder().encode(['address', 'uint256'], [l2User.address, fee])
+    await waitToRelayMessageToL1(
+      l2DAITokenBridge.connect(l2User).withdrawTo(l2Dai.address, lp.address, depositAmount, defaultGasLimit, extraData, ZERO_GAS_OPTS),
+      watcher,
+    )
+
+    // check post-withdrawal state
+    const userL2Balance = await l2Dai.balanceOf(l2User.address)
+    expect(userL2Balance.toString()).to.be.eq('0')
+    const lpL1Balance = await l1Dai.balanceOf(lp.address)
+    expect(lpL1Balance.toString()).to.be.eq(depositAmount)
+    const wasWithdrawn = await lp.wasWithdrawn(l1Dai.address, l2Dai.address, l1User.address, l2User.address, depositAmount, fee, messageNonce)
+    expect(wasWithdrawn).to.be.true
+
+    return messageNonce
+  }
+
+  it('allows LP to claim funds after successful fast withdrawal', async () => {
+    await deposit()
```

Comment:
it's not entirely clear for me why these tests always process slow withdrawal anyway?